VMEncryption/main/check_util.py

```diff
@@ -90,7 +94,83 @@ def is_invalid_lvm_os(self):
                     detected = True
         return detected
 
-    def is_precheck_failure(self):
+    def check_kv_url(self, test_url, message):
+        """basic sanity check of the key vault url"""
+        expected = "https://{keyvault-name}.vault.azure.net/"
+
+        if test_url is None:
+            raise Exception(message + '\nNo URL supplied')
+
+        try:
+            parse_result = urlparse.urlparse(test_url)
+        except:
+            raise Exception(message + '\nMalformed URL: ' + test_url)
+
+        if not (parse_result.scheme.lower() == "https" and \
+                re.match('[a-zA-Z0-9\-]+\.vault\.azure\.net(:443)?$', parse_result.netloc)):
+            raise Exception('\n' + message + '\nActual: ' + test_url + '\nExpected: ' + expected + "\n")
+        return
+
+    def check_kv_id(self, test_id, message):
+        """basic sanity check of the key vault id"""
+        # more strict checking would validate the full key vault id format
+        expected = "/subscriptions/{subid}/resourceGroups/{rgname}/providers/Microsoft.KeyVault/vaults/{vaultname}"
+
+        if test_id is None:
+            raise Exception(message + '\nNo Resource ID supplied')
+
+        id_splits = test_id.lower().split('/')
+
+        if not (len(id_splits) >= 9 and \
+                id_splits[0] == "" and \
+                id_splits[1] == "subscriptions" and \
+                id_splits[2] != "" and \
+                id_splits[3] == "resourcegroups" and \
+                id_splits[4] != "" and \
+                id_splits[5] == "providers" and \
+                id_splits[6] == "microsoft.keyvault" and \
+                id_splits[7] == "vaults" and \
+                id_splits[8] != ""):
+            raise Exception('\n' + message + '\nActual: ' + test_id + '\nExpected: ' + expected + "\n")
+        return
+
+    def validate_key_vault_params(self, public_settings):
+        kek_url = public_settings.get(CommonVariables.KeyEncryptionKeyURLKey)
+        kv_url = public_settings.get(CommonVariables.KeyVaultURLKey)
+        kv_id = public_settings.get(CommonVariables.KeyVaultResourceIdKey)
+        kek_kv_id = public_settings.get(CommonVariables.KekVaultResourceIdKey)
+        kek_algorithm = public_settings.get(CommonVariables.KeyEncryptionAlgorithmKey)
+
+        self.check_kv_url(kv_url, "Key Vault URL is required, but was missing or invalid")
+        self.check_kv_id(kv_id, "Key Vault ID is required, but was missing or invalid")
+        if kek_url:
+            self.check_kv_url(kek_url, "A KEK URL was specified, but was invalid")
+            self.check_kv_id(kek_kv_id, "A KEK URL was specified, but its KeyVault ID was invalid")
+            if kek_algorithm not in CommonVariables.encryption_algorithms:
```

Comment:
This will enforce case sensitive check of the algorithm name (the whitelist consists of all upper case strings).  If KV is not case sensitive, then this might reject values that were used before and previously considered valid.

Reply:
Object identifiers are case-insensitive but the documentation doesn't specify that for algorithm identifiers.  Will check with KV team on this  https://docs.microsoft.com/en-us/azure/key-vault/about-keys-secrets-and-certificates#objects-identifiers-and-versioning

Reply:
KV algorithm identifiers are case insensitive (upper and lower case allowed).  Check now supports this, so we are good.  https://github.com/Azure/azure-linux-extensions/pull/676/files#diff-925b65e5f426b09d40462e43eb3b4a82R169